File: scripts/validate.ts
#!/usr/bin/env tsx
/**
 * CLI validator — `npm run validate`.
 *
 * Validates every content/*.json against curriculum.schema.json (via ajv) and
 * checks that all prerequisite ids resolve. Fails (exit 1) on any problem so it
 * can guard the build and run as a pre-commit hook. TypeScript port of the
 * reference scripts/validate.py, sharing the app's own validation module.
 */
import { readFileSync, readdirSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { validateCorpus } from '../src/lib/validation.ts';

const ROOT = dirname(dirname(fileURLToPath(import.meta.url)));
const SCHEMA_PATH = join(ROOT, 'curriculum.schema.json');
const CONTENT_DIR = join(ROOT, 'content');

function main(): void {
  const schema = JSON.parse(readFileSync(SCHEMA_PATH, 'utf8'));

  const fileNames = readdirSync(CONTENT_DIR)
    .filter((f) => f.endsWith('.json'))
    .sort();

  if (fileNames.length === 0) {
    console.error('No content files found.');
    process.exit(1);
  }

  // Parse each file; a JSON syntax error is itself a validation failure.
  const files: Record<string, unknown> = {};
  let parseErrors = 0;
  for (const name of fileNames) {
    try {
      files[name] = JSON.parse(readFileSync(join(CONTENT_DIR, name), 'utf8'));
    } catch (err) {
      parseErrors++;
      console.error(`INVALID JSON in ${name}: ${(err as Error).message}`);
    }
  }

  const result = validateCorpus(schema, files);

  for (const issue of result.issues) {
    console.error(`${issue.source}: ${issue.message}`);
  }

  const projects = Object.values(result.projects);
  const lessonCount = projects.reduce((n, p) => n + p.lessons.length, 0);
  const bands: Record<string, number> = {};
  for (const p of projects) bands[p.age_band] = (bands[p.age_band] ?? 0) + 1;

  console.log(`\n${projects.length} projects, ${lessonCount} lessons`);
  for (const b of ['9-10', '11-12', '13-14', '15-16']) {
    if (bands[b]) console.log(`  ${b}: ${bands[b]}/16`);
  }

  if (parseErrors > 0 || !result.ok) {
    console.error(`\nFAILED: ${parseErrors + result.issues.length} problem(s). No content may enter the app.`);
    process.exit(1);
  }

  console.log('\nAll valid. Prereqs resolve.');
}

main();

File: scripts/validate.py
#!/usr/bin/env python3
"""Validate every content file against curriculum.schema.json and check that all
prerequisite IDs resolve. Run from the repo root:  python scripts/validate.py
Requires: pip install jsonschema
"""
import json, sys, glob, os

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
schema_path = os.path.join(ROOT, "curriculum.schema.json")
content_glob = os.path.join(ROOT, "content", "*.json")

try:
    import jsonschema
except ImportError:
    sys.exit("Please: pip install jsonschema")

schema = json.load(open(schema_path))
files = sorted(glob.glob(content_glob))
if not files:
    sys.exit("No content files found.")

# 1) JSON + schema validation
projects = {}
for f in files:
    name = os.path.basename(f)
    try:
        data = json.load(open(f))
    except json.JSONDecodeError as e:
        sys.exit(f"INVALID JSON in {name}: {e}")
    try:
        jsonschema.validate(instance=data, schema=schema)
    except jsonschema.ValidationError as e:
        sys.exit(f"SCHEMA ERROR in {name}: {e.message}")
    projects[data["id"]] = data

# 2) Prerequisite resolution
errors = 0
for pid, d in projects.items():
    for p in d.get("prereq_project_ids", []):
        if p not in projects:
            print(f"MISSING PREREQ: {pid} -> {p}")
            errors += 1
    lesson_ids = {l["id"] for l in d["lessons"]}
    for l in d["lessons"]:
        for lp in l.get("prereq_lesson_ids", []):
            if lp not in lesson_ids:
                print(f"MISSING LESSON PREREQ: {l['id']} -> {lp}")
                errors += 1

lessons = sum(len(d["lessons"]) for d in projects.values())
bands = {}
for d in projects.values():
    bands.setdefault(d["age_band"], 0)
    bands[d["age_band"]] += 1

print(f"\n{len(projects)} projects, {lessons} lessons")
for b in ["9-10", "11-12", "13-14", "15-16"]:
    if b in bands:
        print(f"  {b}: {bands[b]}/16")
if errors:
    sys.exit(f"\n{errors} prerequisite error(s).")
print("\nAll valid. Prereqs resolve.")
